hashicorp/terraform
# website: TF-153: Split core Terraform docs into "Language" and "CLI"

This PR is the first (and most dramatic) portion of the work described in RFC TF-153.

## What's Up

I'm smashing apart the current Terraform CLI docs into two new top-level docs categories: the Terraform Language Docs and the Terraform CLI Docs. TF-153 goes into much deeper detail on the reasoning, but the short version is:

- The CLI docs are specifically about accomplishing Terraform workflows using the `terraform` binary.
- The language docs are about creating and modifying Terraform configurations, which is relevant to EVERY user of Terraform (even a Terraform Cloud user who never actually installs the `terraform` binary).
- Splitting the docs like that should make everything much more knowable and navigable.

So here's what's inside this PR:

- **Nav changes:** Add a `language.erb` layout, and move the Terraform Language docs there (from out of `docs.erb`). Repurpose `docs.erb` as the Terraform CLI docs. Absorb pages from all of the "island" sidebar navs, so the two new navs are totally consolidated.
- **New content:** Add about 20 new "glue" / "overview" pages to help provide context to the new docs structure. These are mostly small, but it ends up being a decent chunk of text.
- **Smaller changes:**
    - Destroy the "Guides" category and move its content elsewhere.
    - Rename the "Registry" category to "Publishing Providers & Modules".

It's going to be easiest to review this PR by commits, instead of as a whole. The first two commits are just like 230 one-liner changes to adopt pages into new sidebars, and those will get in the way of everything if you don't filter them out.

## Meta-Goals for this PR

There are a few things about this PR that might seem odd, but I did them because I was working under the following constraints:

- Break zero existing links, and limit the splash damage from content changes. Since this is the most dramatic part of the split-up work, I tried to make it highly resistent to PR rot and merge conflicts. (The upcoming PRs are the really nasty part, which is why I'll be keeping them smaller. For the big-bang PR, I wanted it to be as safe as possible.)
    - In a few cases, this means I made some obsolete pages into "ghosts" (still present and viewable, but can't be reached via any navigation elements) rather than delete and redirect them. That's still important to do, but it can't be a blocker on this PR.
    - This also means that the logical navigation hierarchy (the set of "folders" you browse in the sidebar) is now thoroughly out of step with the "physical" URL hierarchy. That's mildly gross if you're the type of person who looks at the URL bar (like... me...), but for the time being, it's better than the alternative. (To be clear: the alternative is a world-ending wave of redirects and link updates that touches every single page of the Terraform documentation. It would stall this project approximately forever.)
        - We'll likely need to rearrange files extensively when transitioning all content to the new web platform (which is blocked on ejecting the remaining provider docs, and it'll likely be Some Months from now). At _that_ time, it'll make sense to finally rationalize this URL hierarchy.
    - This is also why I made one new nav sidebar (language) but kept `docs.erb` as the CLI docs sidebar -- fewer pages to change.
    - This also resulted in some duplicate content in the provisioners docs. Will deal with that later, as part of the follow-up PRs.
- When adding completely new pages, put them in the spot we'd put them if we were building the docs from scratch today. (/docs/cli and places like /docs/configuration/blocks/...) By not trampling on existing URLs, this made it easier to avoid breaking links.
- Make the sidebar navigation behavior match our product sites on the new web platform, like Consul and Nomad et al. That is:
    - "Folders" are always just folders, not "folder and also a page." Clicking a nav item with a disclosure triangle by it will open the folder to reveal contents without navigating you away from the current page.
    - Nearly all folders start with an "overview" page that provides context for their other contents. Exceptions are usually deeply nested.
- Unify sidebars to reduce "islands" and disorientation. That means bringing in all the stuff that ended up in miniature sub-navs, like the function references or the sub-sub-command pages (`terraform workspace list` et al.), as members of _one_ sidebar for the entire docs category. This can result in a little bit more scrolling, but not _too_ much more, and in exchange it makes the docs much easier to navigate and understand.

## Release Plan

- After this is merged to the main development branch, I'll backport it to v0.14 (not v0.13, unless that proves MUCH easier than expected). We'll publish these changes with the 0.14.0 release.
- These changes depend on some minor changes to hashicorp/terraform-website, which are available in https://github.com/hashicorp/terraform-website/tree/oct20_support_language_and_cli_docs. Some (the symlinks) must be merged prior to release, and some (nav changes, deleting old symlinks) must be merged afterwards.
- I plan to make one or more additional PRs of TF-153 work, to split up the "Expressions," "Resources," and "Module Calls" pages into sections. (This PR anticipates that work and leaves room for it, but is mostly independent of it.)

## Testing

To browse a local build of this PR, you'll need to:

- Check out [this terraform-website branch](https://github.com/hashicorp/terraform-website/tree/oct20_support_language_and_cli_docs).
- Make sure the `ext/terraform` submodule is initted, then go into the submodule and `git fetch origin pull/26723/head && git checkout FETCH_HEAD`.
- Return to the top of terraform-website and run `make website`.
